pyuvdata/uvcal/uvcal.py

```diff
@@ -3351,6 +3356,436 @@ def _convert_to_filetype(self, filetype):
             setattr(other_obj, p, param)
         return other_obj
 
+    def initialize_from_uvdata(
+        self,
+        uvdata,
+        gain_convention,
+        cal_style,
+        future_array_shapes=True,
+        metadata_only=True,
+        include_uvdata_history=True,
+        cal_type="gain",
+        times=None,
+        integration_time=None,
+        time_range=None,
+        frequencies=None,
+        channel_width=None,
+        flex_spw=None,
+        flex_spw_id_array=None,
+        wide_band=None,
+        freq_range=None,
+        spw_array=None,
+        jones=None,
+        ref_antenna_name=None,
+        sky_catalog=None,
+        sky_field=None,
+        diffuse_model=None,
+        baseline_range=None,
+        Nsources=None,  # noqa
+        observer=None,
+        gain_scale=None,
+        git_hash_cal=None,
+        git_origin_cal=None,
+        extra_keywords=None,
+    ):
+        """
+        Initialize this object based on a UVData object.
+
+        Parameters
+        ----------
+        uvdata : UVData object
+            The UVData object to initialize from.
+        gain_convention : str
+            What gain convention the UVCal object should be initialized to
+            ("minus" or "plus").
+        cal_style : str
+            What calibration style the UVCal object should be initialized to
+            ("sky" or "redundant").
+        future_array_shapes : bool
+            Option to use the future array shapes (see `use_future_array_shapes`
+            for details).
+        metadata_only : bool
+            Option to only initialize the metadata. If False, this method also
+            initializes the data-like arrays with the appropriate sizes.
+        include_uvdata_history : bool
+            Option to include the history from the uvdata object in the uvcal history.
+        cal_type : str
+            What cal_type the UVCal object should be initialized to
+            ("gain", or "delay").
+        times : array_like of float, optional
+            Calibration times in decimal Julian date. If None, use all unique times from
```

Comment:
Hmm. I'm a little hesitant to make this so specifically tuned to BDA data. You can always pass in a time array that does exactly what you want. If we want this to have special defaults for BDA data I think I need to add another keyword to let users indicate that.

Reply:
I guess we could test to see if the integration time was varying -- that would indicate that there might be BDA (or some other screwiness).

Reply:
We discussed this on our telecon today. We think that allowing the user to pass in the times they want make this flexible enough and don't want to add more options to allow for any possible desired thing with BDA. In our discussion we agreed that using the `integration_time` array to detect baselines with different levels of averaging would be the better approach for the HERA case, so you might want to do that instead of what you showed in that link.